### ISSUE TITLE
mpv-rpi cannot start on raspberry pi 4

### ISSUE THREAD

**Issue by grebenyukaa** [mpv-player/mpv#7492]
### Important Information

Provide following Information:
- mpv version
`mpv 0.30.0 Copyright © 2000-2019 mpv/MPlayer/mplayer2 projects
 built on Thu Feb 27 11:09:22 MSK 2020
ffmpeg library versions:
   libavutil       56.31.100
   libavcodec      58.54.100
   libavformat     58.29.100
   libswscale      5.5.100
   libavfilter     7.57.100
   libswresample   3.5.100
ffmpeg version: n4.2.1`

- Linux Distribution and Version
Archlinux ARM 32bit
`Linux hostname 4.19.105-1-ARCH #1 SMP PREEMPT Sun Feb 23 20:23:32 UTC 2020 armv7l GNU/Linux`

- Source of the mpv binary
Built from source with the help of [https://aur.archlinux.org/packages/mpv-rpi/](url) aur package

- If known which version of mpv introduced the problem
not applicable to particular version

- Window Manager and version
not applicable to particular window manager

- GPU driver and version
vc4-fkms-v3d

- Possible screenshot or video of visual glitches
no screenshot, just this error message:
`➜ ~ mpv --vo=gpu --hwdec=mmal 
movie.mkv`
 `(+) Video --vid=1 (*) (h264 1920x1080 24.000fps)`
 `(+) Audio --aid=1 --alang=eng (*) (eac3 6ch 48000Hz)`
`(+) Subs  --sid=1 --slang=rus (*) 'TeamNotDead' (subrip)`
`     Subs  --sid=2 --slang=eng (subrip)`
`* failed to add service - already in use?`

### Reproduction steps

- Install mpv-rpi with dependencies.
- Try to run it either headless or from an X session. 

The result does not change if you specify --vo=rpi, nor does it change if you specify no hardware decoding.

I already inquired about the nature of this error on the pi forums, here is the answer 
 >mpv is trying to use the old (legacy) firmware GLES driver which does not exist on the Pi4.
Source: [https://www.raspberrypi.org/forums/viewtopic.php?f=28&t=266123&p=1617098#p1617098](url)

I already checked, if mpv is using a correct EGL binary, as I have mesa & libglvnd installed - it is trying to use the vendor blobs, not mesa. Verified with strace:
>`➜  ~ strace mpv --vo=gpu --hwdec=mmal movie.mkv 2>log`
>`openat(AT_FDCWD, "/opt/vc/lib/libEGL.so", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = 3`
>`openat(AT_FDCWD, "/opt/vc/lib/libbrcmEGL.so", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = 3`
>`openat(AT_FDCWD, "/opt/vc/lib/libbrcmGLESv2.so", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = 3`

### Expected behavior
mpv runs normally

### Actual behavior
mpv reports an error and exits.

### Log file

https://pastebin.com/QhMT9LeF

### Sample files
none

**Comment by wm4** [comment]
The native video output (vo_rpi, or vo_gpu with mmal overlay) requires the vendor blobs. This stuff bitrotted with newer distros, and now everything is broken. Also I don't care about this device (why use a failed settop-box SoC with tons of required and extremely cumbersome closed firmware blobs?). Patches welcome, that is all.

**Comment by grebenyukaa** [comment]
Well, the mpv, built without --enable-rpi & --enable-rpi-mmal runs normally, when invoked with --vo=gpu (from an X session). However no hardware decoding is available an software decoding is not very performance-friendly =(

And mpv --vo=gpu --gpu_context=drm also does not want to run on pi4, but with another error.

So the only option to run mpv on this device is to launch the default build without pi-related flags like this: startx mpv --vo=gpu, which does a lot of unnecessary work.

Anyway, I'm not really familiar with those EGL/GLES/whatever there also exists gl-related. Is mpv really trying to use some old GLES driver for rpi video out, as they say?

**Comment by wm4** [comment]
Yes, using the closed vendor blobs is the only way to get hw decoding (at least mmal is actually open on the client side, but that doesn't help much).
I heard building mpv without rpi support and then using mmal-copy works somewhat. (But still not as good as with native display.)
I wonder if rpi will eventually support v4l decoding & drm display.

**Comment by 6by9** [comment]
You have both V4L2 stateful decode (H264 on Pi4, H264/MPEG4/H263 and optionally MPEG2 on earlier boards) and DRM/KMS on the Pi. DRM/KMS is now the default rendering API on Pi4, and optional on Pi0-3.

GL/GLES is via the standard Mesa libraries.

It should look pretty much like any other standard Linux system now, just use dmabufs to pass the images around to avoid hammering SDRAM.

**Comment by wm4** [comment]
We have a DRM overlay in vo_gpu, and v4l support in ffmpeg. No ideas how exactly transparency works. Someone just has to test it and fix whatever problems prevent it from working.

**Comment by grebenyukaa** [comment]
If you can tell me how to test it I can try

**Comment by 6by9** [comment]
Sorry, I don't know the detail of your project. What ffmpeg release do you build? Until fairly recently the ffmpeg V4L2 support assumed the world was NV12 only and fell over in a heap for anything else. https://github.com/FFmpeg/FFmpeg/commit/d61cf1b1ebc2477749d7d7825a072400ed24af9f

If you're not using DRM_PRIME as the output from ffmpeg then you will be memcpying every single image frame, and performance will suck. You may get away with it, but you'll be hammering SDRAM for no reason.

**Comment by wm4** [comment]
Here's the overlay code; it does seem to use AV_PIX_FMT_DRM_PRIME:

https://github.com/mpv-player/mpv/blob/master/video/out/opengl/hwdec_drmprime_drm.c

No idea if it requires nv12, probably not.

This only works with `--vo=gpu`, and not the `drm` VO. As a consequence you need `--gpu-context=drm` (or was it something else?). This is the DRM/GBM backend, so the drivers need to provide GBM. You also probably need to select the correct ``--hwdec``. Not sure which one this is, I only see `v4l2m2m-copy`, which would copy back to main RAM. You could check ``--hwdec=help``. The list is produced by enumerating the provided libavcodec methods with `avcodec_get_hw_config()`.

I have no idea whether the overlay really works. For example, the GL surface must be transparent, or you wouldn't see the video. No idea.

**Comment by grebenyukaa** [comment]
Plain mpv without rpi flags, launched with --vo=gpu --gpu-context=drm, does not work with the error:
>`[vo/gpu/opengl] Initializing GPU context 'drm'`
>`[vo/gpu/opengl] Initializing KMS`
>`[vo/gpu/opengl] Device name: /dev/dri/card0`
>`[vo/gpu/opengl] Driver: v3d 1.0.0 (20180419)`
>`[vo/gpu/opengl] Cannot retrieve DRM resources: Invalid argument`
>`[vo/gpu/opengl] Failed to create KMS.`
>`[vo/gpu] Failed initializing any suitable GPU context!`

Here is the full log:
https://pastebin.com/9PKfi9Bj

So it already does not work out of the box.

**Comment by wm4** [comment]
You may need to run this as root.

**Comment by 6by9** [comment]
If I can get mpv to build then I'll have a look at why it is objecting.
Note that DRM requires that you are NOT running X or similar.\

The render side and the GL side are separate drivers on Pi4 - v3d (typically /dev/dri/card0) provides 3D, whilst vc4 (typically /dev/dri/card1) provides rendering. The log reports
```
[vo/gpu/opengl] Initializing GPU context 'drm'
[vo/gpu/opengl] Initializing KMS
[vo/gpu/opengl] Device name: /dev/dri/card0
[vo/gpu/opengl] Driver: v3d 1.0.0 (20180419)
```
which implies it is trying to render through v3d.
Is there a simple parameter to tell the GPU context to use an alternate device node? The nodes do advertise correctly whether they support rendering or not, although I can't recall the relevant info off the top of my head.

**Comment by grebenyukaa** [comment]
Okay, I found one, it's --drm-connector, but it requires you to specify card & output device sumultaneously.
Like so: --drm-connector=1.HDMI-A-1

**Comment by grebenyukaa** [comment]
So, the good news are:
 - I finally managed to get a DRM playback, with `--vo=gpu --gpu-context=drm --drm-connector=1.HDMI-A-1`
 - `--hwdec=mmal-copy` reduces CPU total consumption from 50% to 25% (all cores are at 25% cpu)

However, the bad news are:
 - `--hwdec=mmal-copy` messes up sound. I can very clearly hear various rustling, jitter, even echoes. This doesn't seem to happen in other modes. If it can be helpful, the log is here: https://pastebin.com/muiEs1bm
 - `--hwdec=v4l2m2m-copy` shows no video, while actively trying to do so, the log is here https://pastebin.com/1EQGeD5c

**Comment by grebenyukaa** [comment]
Ok, sound problems disappear if I specify --audio-device=alsa/sysdefault:CARD=ALSA instead of letting pulseaudio do its job.

**Comment by wm4** [comment]
v4l2 decodes at least one frame, but messes up quite badly: `[vd] No video PTS! Making something up. Using 24.000000 FPS.`. It loses the video frame timestamp; either a ffmpeg or driver problem.

Then it crashes with this: `mpv: ../video/out/opengl/utils.c:87: gl_upload_tex: Assertion `stride > 0' failed.` This probably means stride is 0, which is clearly ffmpeg violating its own API. (Negative strides are handled before the assert().)

**Comment by 6by9** [comment]
```Assertion stride > 0' failed``` is almost certainly the NV12 only support in FFmpeg. It defines an YU12 format but fails to set the stride and ptr/offset on the chroma planes, leaving them at 0.

**Comment by wm4** [comment]
Do you mean the ffmpeg change you pointed out earlier? It looks like it should be included in the build that triggered the assert. (But not sure. Anyway, the change is pretty old by now.)

**Comment by 6by9** [comment]
Yes, that commit. It's why I asked which version of FFMpeg you use. I wrote it in 2018, but it didn't get merged upstream until Sept 2019
```
commit d61cf1b1ebc2477749d7d7825a072400ed24af9f
Author:     Dave Stevenson <dave.stevenson@raspberrypi.org>
AuthorDate: Thu Mar 22 16:01:35 2018 +0000
Commit:     Aman Gupta <aman@tmm1.net>
CommitDate: Tue Sep 10 18:56:40 2019 -0700

    avcodec/v4l2_buffers: Add handling for NV21 and YUV420P
    
    The single planar support was for NV12 only.
    Add NV21 and YUV420P support.
    
    Signed-off-by: Aman Gupta <aman@tmm1.net>
```
Someone pointed out the other day that it wasn't in the 4.2 release.

**Comment by wm4** [comment]
Oh yeah, then it's possible that it was missing from that one. (I only looked at which date the user's ffmpeg was built, according to the release number in the mpv log file.)

**Comment by grebenyukaa** [comment]
@6by9
`ffmpeg version n4.2.1 Copyright (c) 2000-2019 the FFmpeg developers`

It came from ffmpeg-mmal package like this. I guess I can build a more contemporary version, there is a ffmpeg-git package, which will need just one more flag to enable mmal.

**Comment by grebenyukaa** [comment]
With ffmpeg version `ffmpeg version N-96858-g516e9b9d72 Copyright (c) 2000-2020 the FFmpeg developers`, which is the current master https://github.com/FFmpeg/FFmpeg/commit/516e9b9d72be1698bc232db652dad19389a16db5, both modes are working.

I see no major differences in playback quality with either `v4l2m2m-copy` or `mmal-copy` hw decoding.

Performance-wise both modes consume nearly the same amount of total CPU - `mmal` one loads all 4 cores evenly at about 20%, `v4l2` one loads most of the time only 1 core by 80% and the others by ~0.

Personally I find v4l2 one more smooth, but that might just be placebo.

**Comment by 6by9** [comment]
I know the LibreElec folk are trying to get FFmpeg DRM_PRIME support fixed for V4L2 stateful decoder API, but I don't think it is there yet. That should allow you to get rid of that copy.

**Comment by grebenyukaa** [comment]
@wm4 could you please fix DRM device auto discovery by mpv? Right now it chooses v3d over vc4 automatically.

Or could you for example implement an option like `--drm-device` or `--drm-gpu`, that doesn't require to specify the name of the connector, but allows to set the number of the gpu or the name of the gpu device?

It's not very robust or convenient to specify `--drm-connector=1.HDMI-A-1` every time, and it will break if I reconnect it to the other hdmi port.

**Comment by wm4** [comment]
I'm not going to do anything (other than removing the vendor-blob specific RPI code from mpv when it's finally time), I'm happy with x11 on desktop.
Maybe @xantoz is interested.

**Comment by grebenyukaa** [comment]
I doubt that it is platform dependent. All vendor blobs aside, your algorithm for drm device automatic discovery selected a card with no connected screens and drm resources among others.

```
> ~ mpv --drm-connector=help
Available connectors for card 0:
Cannot retrieve DRM resources: Invalid argument

Available connectors for card 1:
  HDMI-A-1 (connected)
```

So it might be worth fixing to further improve usability and user-friendliness

**Comment by szwenni** [comment]
Hi,
So I also tried playbacking h256 files with raspbian.
First of all I think with the standard kernel and the standard ffmpeg and mpv it is not possible in an easy "achievable" way.
I started with trIng tobcompile the kodi_v4l2_1 ffmpeg from popcornmix (https://github.com/popcornmix/FFmpeg/tree/kodi_v4l2_1).
As the v4l2_requests specific stuff ist not available in 4.19.y kernel I was looking for newer kernels and found one from popcornmix again which seems to have the necessary code (https://github.com/popcornmix/linux/tree/v4l2_3).

 After compiling ther kernel and activating the `hevc_codec.dtbo` overlay the rpivid driver starts up with the hevc decoder. Afterwards I patched out the checks for v4l2-request in configure script from ffmpeg. The mandatory code is in kernel and ffmpeg but not in public api of the kernel so the checks for the fourcc slice formats failes there but its irrelevant for now.

So after compiling/installing kernel (zImage, modules and headers!) and ffmpeg from popcornmix I finally had a working ffmpeg with the code to interface the rpivid driver (v4l2-requests).


As I thought it might be good to have the latest mesa gles driver I compiled and installed the mesa master from git. I don't know if this step is really necessary but it should bring performance boost.

Now to MPV:
MPV can be used with the v4l2-request (`--hwdec=drm`) out of the box. Together `--vo=gpu --gpu-context=drm --drm-connector=...` it actually outputs something but nothing useful (Green and pink lines in multiple squares). Looking at the code I figured out that rpivids output pixel format is `V4L2_PIX_FMT_SAND8`. Looking at ffmpeg v4l2-request code at [this line](https://github.com/popcornmix/FFmpeg/blob/99dc9cec7cde717c58db8dd0a2e947644785a06a/libavcodec/v4l2_request.c#L182) you could see that a modifier is set. Looking at mpvs `drm-prime.c` I saw that the modifiers are not forwarded to kernel via libdrm in `drm_prime.c`'s  function`drm_prime_create_framebuffer`. Instead of using `drmModeAddFB2` `drmModeFB2WithModifiers` should be used here. Modifiers are `uint64_t [4]`. For all used planes modifiers must be the same and are refering from `descriptor->objects[0].format_modifier`. Looking at `vc4_firmware_kms.c` from kernel the modifiers leads to an image format switch (makes sense as the decoder uses an non-standard yuv output [see here](https://github.com/popcornmix/linux/blob/269c081c2fa6e2b8f1361b9f483a0d7aa709e23a/drivers/gpu/drm/vc4/vc4_firmware_kms.c#L631)).

Now the ouput looks good, but on one of my videos I tested it starts flickering quite heavy. The video itself looks fine but there are white lines over the video. I can post a picture tomorrow. An other video worked perfectly (https://s3.amazonaws.com/x265.org/video/BigBuckBunny_2000hevc.mp4). Both H265 4k 8 bit. The only difference is the framerate. The flickering one has 29.97 fps the working one 25fps.

So now to the main question, is this related to mpv ? Maybe destroying the drm frames to early ? Or is it an DRM/KMS problem.

Maybe @6by9 or @popcornmix can say anything to the flickering problem if its possible that this is related to vc4 drm code or the v4l2 decoding.

@wm4 is it possible that this is because of the framerates ?

If somebody wants the ffmpeg config and mpv config I can post them later when I have my pc back.

I also found an other bug in `context_drm_egl.c` which is also called with the above parameters (I don't get why). In [this line](https://github.com/mpv-player/mpv/blob/e7bab0025f3617a2e19a1b9eacf689423787ed89/video/out/opengl/context_drm_egl.c#L276) it makes an invalid request. I will debug this tomorrow to get more information.

Long story short:
The good news are that mpv is able to play videos on rpi4 without any broadcom binaries compiled into it. Only a little modification is needed in `drm_prime.c`. Bad news are until all the code is available in mainline/downstream kernel it will take some time.

**Comment by wm4** [comment]
> @wm4 is it possible that this is because of the framerates ?

Don't ask me, I know nothing about V4L or DRM.

Can you send patches for the things you corrected?

> Bad news are until all the code is available in mainline/downstream kernel it will take some time.

Isn't that the usual RPI situation?

**Comment by szwenni** [comment]
> Don't ask me, I know nothing about V4L or DRM.

Who created the drm prime part of mpv ?

> Can you send patches for the things you corrected?

I will create an pull request tomorrow with the patch.

> Isn't that the usual RPI situation?

Seems like to much effort for an easy seamless looping playback of a video without having kodi libreelec or something.

**Comment by wm4** [comment]
> Who created the drm prime part of mpv ?

That was @LongChair, but he's not active anymore I guess.

**Comment by 6by9** [comment]
V4L2 HEVC stateless decoder API hasn't been finalised in mainline, therefore anything working on HEVC is still in flux.
I'm in the process of making a clean 5.4 tree with all relevant backports and patches for our HEVC block, and it'll also be sent to linux-media to look at mainlining it.

We know there is a current issue with the vc4-fkms-v3d (hybrid kernel/firmware) driver and when it is reporting flip complete events. That can result in image tearing.
The first revision of patches to provide a kernel side only driver went to dri-devel last week. (The work has been contracted out to Maxime Ripard).

**Comment by szwenni** [comment]
@wm4 okay yeah thats okay. It seems to work in most cases so no problem. I need to clean up the code then I will create a pull request.

@6by9 yeah I've read that a lot in the comments of the files but as I found the same structures in different code projects (kernel, ffmpeg, libva-backend from bootlin) it seems like it is quite stable.

I have seen the vc4 patches from dri-devel (Maxime) right now. I'm looking forward having them in some upstream repository.

But for now I think there isn't that much I can do anymore for getting the videos to play with HEVC decoding than waiting.

**Comment by wm4** [comment]
Did the kernel folks specify the API anywhere, or is it the usual "reverse engineer the entire Linux source code" again? Most docs about v4l2 seem to be at best abstract or overly generic. In any case, almost everything about this happens in libavcodec anyway.

Is there still any functionality that is covered by the vendor specific APIs only?

**Comment by 6by9** [comment]
In line with the best of them, always do things twice.

V4L2 stateful decoder API  - https://linuxtv.org/downloads/v4l-dvb-apis-new/uapi/v4l/dev-decoder.html.
This is for hardware which do all the bitstream parsing for themselves. Feed in the raw ES (and timestamps), and get frames (and timestamps) out.

V4L2 stateless decoder API - https://linuxtv.org/downloads/v4l-dvb-apis-new/uapi/v4l/dev-stateless-decoder.html.
This is for hardware which requires software to have parsed the headers to split out which reference frames to use and the like. It involves jumping through a load of hoops to use the V4L2 Request API in order for the metadata to be passed with the correct slice of bitstream.


For the Pi, we have a stateful decoder for the older H264 block (and MPEG4/H263/MPEG2/VC-1 on Pi0-3), whilst the HEVC block on the Pi4 will be using the stateless API.

FFmpeg's v4l2m2m module deals with the stateful API, but AFAIK is still missing exporting buffers as AV_PIX_FMT_DRM_PRIME, which means we can't do zerocopy with dmabufs on those buffers. There have been attempts to get it merged, but limited interest/support/assistance from core FFmpeg devs.

I wasn't aware of a stateless API implementation having been merged into FFmpeg as yet. https://github.com/popcornmix/FFmpeg/tree/kodi_v4l2_1 is the tree that we (and LibreElec) are working on to get it working sensibly. There will then be attempts to get it upstreamed.


I believe LibreElec and Kodi are aiming to have gone solely V4L2, GBM, and DRM with the next release. Whether it will quite make it is a different question, but it should be close.

**Comment by wm4** [comment]
That sounds generally very nice. I'm happy to learn that RPI finally got there.

A bit regrettable that the "old" codecs still use the stateful API. Are there any plans to update them? How does it work on the hardware level? When I looked at MMAL, it simply communicated with the closed source "firmware OS" (whatever it's called) via shared memory.

Regarding ffmpeg, just find someone on the ffmpeg IRC channel who will push it for you. (You'll also need to send the patches to the mailing list of course.) I don't think they're opposed to this new API, if anything, the stateful API had less chances to be accepted, merely going by technical reasons.

**Comment by szwenni** [comment]
Hi,
I think I found the other issue.
Context:
I have 3 Videos in HEVC codec. Two with 1920x1080 (A, B) and two with 3840x2160 (C, D). A, C has a framerate of 25fps B and D 29.97fps. A and C are 10 seconds, B and D 1:30 min. A and C are playing perfectly without any issues. B and D not. B doesn't play at all and D has white stripes in it.

Findings:
MPV allocates GBM dumb buffers which should be backed by cma memory calculated from the height and format strides (in this case ARGB888). Handles are assigned to this buffers by the kernel to get them into userspace.
FFMPEG also allocates buffers which are the drm prime buffers having the video data in it (I think). These are buffers backed by dma memory. These also have handles for referencing them in user space.
Now if I start video B I can see that the that the same handle is used twice. Once coming from ffmpeg and once from GBM. But as ffmpeg already has allocated the buffer as dma memory, its size is too small for putting the ARGB888 data into it. Thats why it fails to create a framebuffer for it in kernel. I can't tell why this happens. I could imagine there is a race condition somewhere in mpv. I can't image it is something kernel related. What I think proves that thesis is that the shorter Video successfully plays and get different handles from kernel for the last GBM buffer and the DRM Prime buffer. This behavior is stable and occurs every time I start that video.

I now also have an explanation why there is this flickering in Video D. There the same behavior occurs. It tries to use the same handle twice. But because of being a 4K video, the buffer is big enough to put the data in even if it isn't the right buffer. As this should be the buffer for the decoded video not for OSD what I think it is.

I can't really tell how to fix this. Maybe egl and gbm has to initialized with all buffers before ffmpeg starts decoding as it seems gbm allocates the buffers when video already started. Maybe this is really a kernel bug or even a mesa bug.

@6by9 has an better overview which bugs are know right now. Is there something similar to this ? If yes this is the wrong place to discuss, but I can send you logging output for this which has some additional logs in the specific kernel modules which cleary shows what happen (but no why).

**Comment by wm4** [comment]
Decoding and display are separate in mpv. The decoder is supposed to allocate its decoding buffers (which are usually something in YCbCr aka YUV), while the display code allocates a framebuffer (usually RGB) for rendering. (Though his prime interop tries to use it as direct hardware overlay. The GL framebuffer is used only for subtitles and OSD.) Decoder and display can share a context, which is supposed to facilitate interop between these parts if it's needed. (For example vaapi, at least without explicitly using "outside" APIs like dmabufs, absolutely requires that decoding and display use the same API context.)

I have no idea what's going on here. You say race condition; maybe that's an unexpected interaction due to the fact that decoding and display run on separate threads asynchronously.

**Comment by szwenni** [comment]
Thats what I understood so far about mpv.
What I don't get is how the gl buffers are actually drawn. But anyway I can't explain why and how this happens and why this happens exactly there. Maybe it is really a bug somewhere in the drm/kms/gbm environment. Maybe worh this big changeset coming to rpi drm environment it will behave normal.

I really don't geht why the one video is working the other one not. The biggest problem is that with debugger attched it "works" like it should. That was an other thing leading me to the assumption of some race condition.

**Comment by szwenni** [comment]
So I investigated again.
The control flow is like e.g. assuming handle number 3:
1. FFMPEG decodes data into drm prime with handle 3
2. The Handle to the drm prime buffer is released (dont know why as FFMPEG should still hold a reference to it)
3. A dumb buffer is created and gets handle 3 as it was recently freed
4. Just before the dumb buffer gets a framebuffer, the drm prime fd to handle inits the buffer again as prime fd is there, the dma buf is still there and handle 3 will reinited with the size from dma buf which is too small for argb data
5. As handle is still there for dumb buffer and has the same number drmmodefbadd tries to create a frame buffer for handle 3 which points to the dma buffer not the cma buffer which it should on this point
6. It fails for obvious reasons

The race here is if the prime is freed just milliseconds later the handle 3 is still there and the dumb buffer will get a different handle
That maybe explains why different frame rates are leading to different results

**Comment by wm4** [comment]
Interesting. FFmpeg refcounts all frames (including those for hardware decoding), so this is very strange.

**Comment by szwenni** [comment]
https://github.com/popcornmix/linux/blob/269c081c2fa6e2b8f1361b9f483a0d7aa709e23a/drivers/gpu/drm/drm_prime.c#L61

Reading this leads to the assumption that userspace must ensure that the handle is unique and the dmabuf is released if the handle is tried to use. This leads to the assumption that there might be a problem with gbm.
But I dont know if this assumption is right. The comment just explains why we are getting the same handle back that "should be available".

**Comment by szwenni** [comment]
Found the bug see #7506

**Comment by wm4** [comment]
So thanks to @szwenni this should work much better now. Is there still anything needed in mpv, other than upstream ffmpeg/kernel changes?

Meanwhile, I think I'll disable the autodetection of the RPI vendor EGL libs, since that seems to be actively harmful.

**Comment by szwenni** [comment]
Just some ideas.
- Try to somehow seperate rpi4 from the others.
- On RPI4 activate v4l2-requests if available in ffmpeg but disable mmal. (I can pass you a list with ffmpeg options) 
- In mpv deactivate all broadcom shit if rpi 4 is detected. (Mesa is there, mmal not needed as we have v4l2m2m(-copy) which is basically mmal just in kernel).
- Maybe add a way to have custom url for ffmpeg.

My thoughts to that topic. Maybe I will do a write up of how to achive this after my bachelor thesis.

**Comment by wm4** [comment]
Is the v4l stuff not available on RPI3, if using the required kernel and ffmpeg patches?
The main problem is really that the broadcom stuff doesn't integrate with Mesa at all.

**Comment by szwenni** [comment]
So v4l2 for h264 and below should be there for rpi3 also even in mainline (? Have to look it up). The v4l2-request is only on rpi4 as it has the hevc hardware dec.

On Rpi4 we dont need any of the broadcom stuff.
We don't need the userland mmal stuff for decoding as kernel can do it on rpi3/4 and h264. We don't need brcmegl as mesa is fully supported on at least rpi4. So on rpi4 I don't see the need of brcm binaries.
Upstream ffmpeg/kernel and mesa is able to do the same.

**Comment by c1ngular** [comment]
how is everything going now ?

**Comment by danboid** [comment]
I'd also like to run mpv with hw accel decoding on RPi 4.

I have got hw accelerated video playback working under Raspbian Buster VLC but non of the suggested mpv commands and options have resulted in accelerated playback for me. From skimming this thread it seems like it is (or was) also an issue with the Raspbian kernel. Has the Raspbian kernel been fixed now?

**Comment by gravity-addiction** [comment]
> Has the Raspbian kernel been fixed now?

Any resolution on this? My jetson nano is working awesome, but i have a pile of raspberry pi's here that are collecting dust until I can use mpv on them again. Wish I could still by 3b+ as they work the better than a pi4 right now.

**Comment by G2G2G2G** [comment]
@gravity-addiction vlc works on pi4.. omx-player doesn't even exist now and mpv is software =/

**Comment by CounterPillow** [comment]
None of the core developers care about rpi, so it's up to rpi users to make the necessary contributions to support their e-waste.

**Comment by cyph84** [comment]
hwdec=mmal only works with vo=mmal. I've recently just tested that on an Rpi 3b+

**Comment by rien333** [comment]
> hwdec=mmal only works with vo=mmal

Think you mistyped something - `vo=mmal` probably never existed?

**Comment by cyph84** [comment]
@szwenni Were you doing your experiments with --hwdec=v4l2m2m-copy?

I am trying to get zero-copy v4lm2m working on the Le Potato and can't seem to get --hwdec=v4l2m2m as a valid hwdec.
